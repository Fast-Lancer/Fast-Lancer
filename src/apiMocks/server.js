import { rest } from 'msw'
import { setupServer } from 'msw/node'

const url = process.env.REACT_APP_SUPABASE_URL + '/rest/v1'

let postedClient = {}
let postedProject = {}

export const server = setupServer(
  rest.get(url + '/clients', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(select === '*' && !id) {
      // response for getClients()
      return res(
        ctx.json([
          {
            id: 1,
            client_name: 'bob1',
            business_name: 'business1'
          },
          {
            id: 2,
            client_name: 'bob2',
            business_name: 'business2'
          },
          {
            id: 3,
            client_name: 'bob3',
            business_name: 'business3'
          }
        ])
      )
    } else if(id) {
      // response for getClient(id)
      if(id === 'eq.42') {
        // Janky workaround to make testing edit/create possible.
        // The post route mock will set the id to 42
        // So this responds with the posted data when client.id is 42
        return res(
          ctx.json(
            postedClient
          )
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              client_name: 'bob1',
              business_name: 'business1'
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/clients', (req, res, ctx) => {
    let [client] = req.body
    client = { ...client, id: 42 }
    postedClient = client
    return res(
      ctx.json([
        client
      ])
    )
  }),
  rest.get(url + '/projects', (req, res, ctx) => {
    const select = req.url.searchParams.get('select')
    const id = req.url.searchParams.get('id')
    if(!id) {
      // response for getProjects()
      return res(
        ctx.json([
          {
            id: 1,
            title: 'project1',
            client_id: '1',
            clients: {
              client_name: 'bob1'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          },
          {
            id: 2,
            title: 'project2',
            client_id: '2',
            clients: {
              client_name: 'bob2'
            },
            
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          
          },
          {
            id: 3,
            title: 'project3',
            client_id: '3',
            clients: {
              client_name: 'bob3'
            },
            date_start:'2022-01-20',
            date_end:'2022-01-21',
            URL: 'something@something.com',
            description: 'ok',
            notes: 'test notes',
            hourly_rate: 33,
            hours_quoted: 22,
            price_quoted: 300
          }
        ])
      )
    } else if(id) {
      // response for getProjectById(id)
      if(id === 'eq.42') {
        return res(
          ctx.json(postedProject)
        )
      } else {
        return res(
          ctx.json(
            {
              id: 1,
              title: 'project1',
              client_id: '1',
              clients: {
                client_name: 'bob1'
              },
              date_start:'2022-01-20',
              date_end:'2022-01-21',
              URL: 'something@something.com',
              description: 'ok',
              notes: 'test notes',
              hourly_rate: 33,
              hours_quoted: 22,
              price_quoted: 300
            }
          )
        )
      }
    }
  }),
  rest.post(url + '/projects', (req, res, ctx) => {
    let project = req.body
    project = { ...project, id: 42 }
    postedProject = project
    return res(
      ctx.json([project])
    )
  })
)
