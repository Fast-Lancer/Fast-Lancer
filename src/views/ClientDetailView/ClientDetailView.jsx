import { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import ClientDetail from '../../components/ClientDetail/ClientDetail.jsx'
import Title from '../../components/Title/Title.jsx'
import { getClient } from '../../services/clients.js'

export default function ClientDetailView() {
  const { id } = useParams()
  const [client, setClient] = useState()
  const [loading, setLoading] = useState(true)

  useEffect(() => {
    getClient(id)
      .then(setClient)
      .catch(e => console.error(e))
      .finally(() => setLoading(false))
  }, [])
  
  return (
    <div>
      {
        loading
          ? <h1>Loading...</h1>
          : <div>
            <Title pageTitle='client detail' pageHeader={client.name} />
            <ClientDetail client={client}/>
          </div>
      }
    </div>
  )
}
