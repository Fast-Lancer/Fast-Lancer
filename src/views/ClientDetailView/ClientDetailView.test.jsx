import { render, screen, waitFor } from '@testing-library/react'
import { Route, Switch } from 'react-router-dom'
import { MemoryRouter } from 'react-router-dom'
import { server } from '../../apiMocks/server.js'
import ClientDetailView from './ClientDetailView.jsx'

beforeAll(() => {
  server.listen()
})

afterAll(() => {
  server.close()
})

it('should render the fetched data', async () => {
  const { container } = render(
    <MemoryRouter initialEntries={['/1']}>
      <Switch>
        <Route path='/:id'>
          <ClientDetailView />
        </Route>
      </Switch>
    </MemoryRouter>
  )

  await screen.findByText('Loading...')

  await screen.findByText('Notes')

  // This bob came from the mocked service
  waitFor(() => screen.findByText(/business1/), {
    timeout: 5000
  })

  expect(container).toMatchSnapshot()
})
