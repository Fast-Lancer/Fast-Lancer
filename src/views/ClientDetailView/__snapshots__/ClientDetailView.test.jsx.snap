// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the fetched data 1`] = `
<div>
  <div>
    <div>
      <div
        class="titleBody"
      >
        <section
          class="title"
        />
        <ul
          class="tabnav"
        >
          <li
            class="active"
          >
            <a
              href="/clients"
            >
              Clients
            </a>
          </li>
          <li>
            <a
              href="/projects"
            >
              Projects
            </a>
          </li>
          <li>
            <a
              href="/editclient"
            >
              Edit Client
            </a>
          </li>
        </ul>
      </div>
      <div
        class="clientDetail"
      >
        <h3
          class="h3"
        />
        <div
          class="twoColRow"
        >
          <div
            class="col"
          >
            <div
              class="row"
            >
              <img
                alt="a phone icon"
                src="phoneSmall.png"
              />
            </div>
            <div
              class="row"
            >
              <img
                alt="an email icon"
                src="emailSmall.png"
              />
            </div>
          </div>
          <div
            class="col"
          >
            <div
              class="row"
            >
              <img
                alt="a palette icon"
                src="paletteSmall.png"
              />
              Client Projects
            </div>
            <ul
              class="linkList"
            >
              <li>
                <a
                  href="/projects/1"
                >
                  project1
                </a>
              </li>
              <li>
                <a
                  href="/projects/2"
                >
                  project2
                </a>
              </li>
              <li>
                <a
                  href="/projects/3"
                >
                  project3
                </a>
              </li>
            </ul>
          </div>
        </div>
        <div>
          <div
            class="row"
          >
            <img
              alt="a notes icon"
              src="emailSmall.png"
            />
            Notes
          </div>
          <div
            class="notesDiv"
          >
            No notes entered.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
